jepsen/src/jepsen/tests/monotonic_cycle.clj

```diff
@@ -0,0 +1,232 @@
+                                                 (recur w stack scc))
+                                               [stack scc]))))
+
+                           ;; SCC good to go sir
+                           sccs (if scc
+                                  (conj sccs scc)
+                                  sccs)
+
+                           ;; Update head of min-stack
+                           min-stack (if (< 0 (count min-stack))
+                                       (let [m         (peek min-stack)
+                                             min-stack (pop min-stack)
+
+                                             i' (aget indices prev-node)
+                                             ;; Find the smaller of the node's index or last min
+                                             low       (min i' m)
+                                             min-stack (conj min-stack low)]
+                                         min-stack))]
+                       (recur index indices visited iterator
+                              iter-stack min-stack stack prev-node sccs)))))]
+    result))
+
+(defn apply-pair
+  "Takes a map of registers to their last-seen values, a map of registers to
+  their graphs, and a tuple containing a register and its new value. Returns
+  the graph-map with the op applied to the appropriate register's graph."
+  [last graph [k v]]
+  (let [last-v (last k)
+        node {v #{}}
+        edge (if last-v
+               {last-v #{v}}
+               {})
+        g (merge-with set/union (graph k) node edge)]
+    (assoc-in graph [k] g)))
+
+(defn graph
+  "Takes a history of reads over a single register and returns a graph of the
+  states that the register advanced through."
+  [history]
+  (loop [graph {}
+         last  {}
+         ;; This will probably have to be a map of keys to last-seen values
+         [op & more] history]
+    (let [val   (:value op)
+          ;; Single register get a key called :register
+          val   (if-not (map? val)
+                  {:register val}
+                  val)
+          pairs (concat val)
+
+          ;; Remove values that have not changed
+          new-pairs (set/difference (set pairs)
+                                    (set last))
+
+          ;; Apply new nodes and edges to the graph
+          g' (reduce (partial apply-pair last)
+                     graph
+                     new-pairs)]
+      (if more
+        (recur g' (merge last val) more)
+        g'))))
+
+(defn errors
+  "Takes a set of component-sets from tarjan's results, identifying if any
+  components are strongly connected (more than 1 element per set)."
+  [components]
+  (let [<=2 (fn [set]
+              {set (<= 2 (count set))})
+        error? (fn [m] (some true? (vals m)))]
+    (->> components
+         (map <=2)
+         (filter error?)
+         (reduce merge))))
+
+(defn valid?
+  "Takes a map of registers to their analyzed components. Returns true if
```

Comment:
I'm a little confused by this, because for a history, we should only be doing doing Tarjan's algorithm once, on a single graph of all operations, rather than once for each register. Is that what's happening here?